lib/3scale_toolbox/configuration.rb

```diff
@@ -0,0 +1,45 @@
+require 'yaml/store'
+
+module ThreeScaleToolbox
+  class Configuration
+    ATTRIBUTES = %i[remotes].freeze
+    private_constant :ATTRIBUTES
+
+    def initialize
+      @store = YAML::Store.new(ThreeScaleToolbox.config_file)
+    end
+
+    def remotes
```

Comment:
This is not very scalable. Every command that wants to store something in its own namespace will need to extend this object.

IMO we should look for better API where each command can put own stuff into the configuration object.

Reply:
I do not like the idea of exposing a hash object so each command can put/get/delete own stuff.  Having class level `ATTRIBUTES` list enables filtering what's can be included in persisted file. 

Instead of making commands extend this object with methods like 
```ruby
def remotes
end 
```
and
```ruby
def update_remotes
end 
```
those methods could be generated dynamically just adding new elements to `ATTRIBUTES` list. The body of the methods should be the same.

Reply:
If I write command as and extension I would not be able to use the config file without monkey patch. That does not sound right to me. In the and it is all Hash structures.

There should be some good reason to write more code than necessary. I really don't see one here. The configuration class can just make sure it opens transaction when running the cli, read / write data and thats it.

I really don't see any benefit of `configuration.remotes` over `configuration.data['remotes']`.

Why would you need to filter / limit what is included in the file? What is the problem we would be solving with this technical solution?

Reply:
I guess this has to do with the typical trade-off flexibility over control. We allow plugins to add stuff into persisted config file and lose control over what is stored vs plugins are not allowed to add stuff (can read it though) and stored content is controlled (at least at first level).

Rubygems does not allow arbitrary content in the config file. All updates are via write accessors like [set_api_key](https://github.com/rubygems/rubygems/blob/master/lib/rubygems/config_file.rb#L302)

Reply:
I don't really see it as a tradeoff. I see it as that there should be some problem to be solved and if there is no problem then there is no reason to write code.

What is the problem we are trying to solve with restricting what can be stored in the config?
That is a serious question. Is it namespaces between commands ?

BTW Rubygems offers raw access: https://github.com/rubygems/rubygems/blob/672ddcbe14f49c48f6c5492b38480a9162680b5b/lib/rubygems/config_file.rb#L447-L455

Reply:
Right, Rubygems offers raw access. 

I guess control over stored content allows you to provide business logic over it, in case needed. 

In your own example, you add control over stored content.
https://github.com/3scale/jenkins-build/blob/master/lib/jenkins/build/configuration.rb

Reply:
My example was done as one day hack to solve a problem and never aimed to be extensible.
It stored configuration just for itself, not for any number of commands written by other people.

If providing some "nicer" access is the issue then the command can implement that by itself  using OpenStruct or something. The business logic on the data can be implemented anywhere.

The responsibility logic of the store should be implemented in the store: transaction, load, save, ...

For me the core question is: will other commands store configuration there?

If the answer is yes, then imagine the situation of 100 commands storing the configuration there. The work that would have to be done to make each of them work.
Then offer the simplest access that fulfills the goal.

If the answer is no, then offer the simplest way for the app to read its config and thats it. No need to be extensible.

Because ultimately there should be just single responsibility. Either being extensible configuration store for data and each command will operate on it how they want or a one instance of this: store for one application with business logic.

You can serialize whatever ruby objects in the YAML store. Each configuration object can have own small config with business logic object stored and serialized.